### PR TITLE
Various ELF loading size checks for safety

### PR DESCRIPTION
**Pull request by hrydgard**
Will hopefully eliminate some of the crashes seen in #9774 .

### CODE REVIEW

File: Core/ELF/ElfReader.cpp [hunk 1]
```diff
@@ -29,8 +29,8 @@ const char *ElfReader::GetSectionName(int section) const {
 		return nullptr;
 
 	int nameOffset = sections[section].sh_name;
-	if (nameOffset < 0) {  // TODO: Where can we get a solid upper limit?
-		ERROR_LOG(LOADER, "ELF: Bad name offset %d in section %d", nameOffset, section);
+	if (nameOffset < 0 || nameOffset >= size_) {
```
Comment (unknownbrackets) [inline Core/ELF/ElfReader.cpp]: Should we also to a `memchr` or something to validate that there is a NUL byte before `size_`?

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]

File: Core/ELF/ElfReader.cpp (hunk 1 above)
Comment (pongtap231) [inline Core/ELF/ElfReader.cpp]: Yes

Comment (hrydgard) [inline Core/ELF/ElfReader.cpp]: Not sure what you mean by before size_, do you mean before nameOffset? And that would only take a byte load, not a memchr.

Comment (unknownbrackets) [inline Core/ELF/ElfReader.cpp]: Sorry, I mean that there's no guarantee a null terminated string exists here.  Let's say that nameOffset is 3:

```
A B C D E F G H I J K L M N O P Q R S T U X X X X X
    ^                         ^
    \-- Start of string       \--- End of ELF
```

No one promised there'd be a NUL byte here anywhere.  We might still, when strcmp()ing with the string, go past size_.

Obviously, if the first byte of `nameOffset` is 0, that'd be fairly stupid.

-[Unknown]

Comment (hrydgard) [inline Core/ELF/ElfReader.cpp]: Ah, I see now. Probably not a common issue, but might be worth checking for indeed.

**Pull request merged**